app/models/pageflow/membership.rb

```diff
@@ -1,9 +1,26 @@
 module Pageflow
   class Membership < ActiveRecord::Base
     belongs_to :user
-    belongs_to :entry
+    belongs_to :entity, polymorphic: true
 
-    validates :user, :entry, :presence => true
-    validates :entry_id, :uniqueness => { :scope => :user_id }
+    validates :user, :entity, :role, presence: true
+    validates :user_id, uniqueness: {scope: [:entity_type, :entity_id]}
+    if :entity_type == 'Pageflow::Entry'
+      validates_with AccountMembershipExistenceValidator
+      validates :role, inclusion: {in: %w(previewer editor publisher manager).map(&:to_sym)}
+    else
+      validates :role, inclusion: {in: %w(member previewer editor publisher manager).map(&:to_sym)}
+    end
+
+    class AccountMembershipExistenceValidator < ActiveModel::Validator
+      def validate(record)
+        unless record.user.accounts.include?(record.entity.account)
+          record.errors[:base] << 'Entry Membership misses presupposed Membership on account of entry'
+        end
+      end
+    end
```

Comment:
If the validator only has one method, we can use [custom method validation](http://guides.rubyonrails.org/active_record_validations.html#custom-methods)